### PR TITLE
:warning: Stop serving v1alpha3 API types

### PR DESCRIPTION
**Pull request by killianmuldoon**
Set `served: false` on v1alpha3 API types in:
-  `api`
- `controlplane/kubeadm/api`
- `booststrap/kubeadm/api`
- `test/infrastructure/docker/api`

Part of #8038

### CODE REVIEW

File: docs/book/src/developer/providers/migrations/v1.4-to-v1.5.md [hunk 1]
```diff
@@ -16,12 +16,14 @@ maintainers of providers and consumers of our Go API.
 ## Changes by Kind
 
 ### Deprecation
+- 
+- API version`v1alpha4` is deprecated and will be removed in v1.6.
```
Comment (fabriziopandini) [inline docs/book/src/developer/providers/migrations/v1.4-to-v1.5.md:20]:
```suggestion
- API version`v1alpha4` is deprecated and CAPI will stop serving this version in v1.6.
```

File: docs/book/src/developer/providers/migrations/v1.4-to-v1.5.md [hunk 2]
```diff
@@ -16,12 +16,14 @@ maintainers of providers and consumers of our Go API.
...
 
--
 
 ### Removals
 
--
+- API version `v1alpha3` is removed from v1.5.
```
Comment (fabriziopandini) [inline docs/book/src/developer/providers/migrations/v1.4-to-v1.5.md:25]:
```suggestion
- API version `v1alpha3` is not served anymore from v1.5 (users can enable it manually in case they are lagging behind with deprecation cycles). Important: `v1alpha3` will be removed in 1.6.
```

**Comment by killianmuldoon** [comment]
>/test pull-cluster-api-e2e-full-main

No need to keep tagging this - I'm currently running the 0.3 migration test in PR-Informing to ensure it's passing.

File: api/v1alpha3/doc.go [hunk 3]
```diff
@@ -18,4 +18,5 @@ limitations under the License.
 // +k8s:conversion-gen=sigs.k8s.io/cluster-api/api/v1beta1
 //
 // Deprecated: This package will be removed in one of the next releases.
+// +kubebuilder:unservedversion
```
Comment (ykakarap) [inline api/v1alpha3/doc.go:21]: Looks like these package level markers are not doing anything. Can we drop them if not needed?

(Same for other packages)

**Comment by sbueringer** [comment]
To avoid confusion, WIP or ready? 😀

**Comment by killianmuldoon** [comment]
Attempted to run the test using a sequential upgrade - `v0.3` -> `v1.4` -> `latest` - but still ran into the deletion issue where the kube-controller-manager is not garbage collecting based on ownerReferences correctly. I'm digging into that issue, but I think a reasonable compromise for now is to make this test upgrade `v0.3` -> `v1.4` only and not test upgrades to a version where `v0.3` is not served.

The end result of this may be that the control plane needs to be rolled out after performing the upgrade of unserved versions.

**Comment by killianmuldoon** [comment]
After digging into this we're hitting this issue: https://github.com/kubernetes/kubernetes/issues/102641

The kube-controller-manager isn't updating the in-memory representation of the child resource. The fix is to restart the kube-controller-manager. We should include a warning about this in our release notes, but I don't think there's a reasonable way to fix this in CAPI.

**Comment by killianmuldoon** [comment]
The error in the garbage collector is:

```
I0518 16:06:32.227891       1 garbagecollector.go:348] error syncing item &garbagecollector.node{identity:garbagecollector.objectReference{OwnerReference:v1.OwnerReference{APIVersion:"cluster.x-k8s.io/v1alpha3", Kind:"Machine", Name:"clusterctl-upgrade-5es5vp-md-0-6b5bd6fbc9-64pzg", UID:"ce9f1e0b-76d3-4767-ac30-06117120d5c8", Controller:(*bool)(nil), BlockOwnerDeletion:(*bool)(nil)}, Namespace:"clusterctl-upgrade"}, dependentsLock:sync.RWMutex{w:sync.Mutex{state:0, sema:0x0}, writerSem:0x0, readerSem:0x0, readerCount:1, readerWait:0}, dependents:map[*garbagecollector.node]struct {}{(*garbagecollector.node)(0xc001aef680):struct {}{}, (*garbagecollector.node)(0xc001aef860):struct {}{}}, deletingDependents:false, deletingDependentsLock:sync.RWMutex{w:sync.Mutex{state:0, sema:0x0}, writerSem:0x0, readerSem:0x0, readerCount:0, readerWait:0}, beingDeleted:false, beingDeletedLock:sync.RWMutex{w:sync.Mutex{state:0, sema:0x0}, writerSem:0x0, readerSem:0x0, readerCount:0, readerWait:0}, virtual:false, virtualLock:sync.RWMutex{w:sync.Mutex{state:0, sema:0x0}, writerSem:0x0, readerSem:0x0, readerCount:0, readerWait:0}, owners:[]v1.OwnerReference{v1.OwnerReference{APIVersion:"cluster.x-k8s.io/v1beta1", Kind:"MachineSet", Name:"clusterctl-upgrade-5es5vp-md-0-6b5bd6fbc9", UID:"b7d0d42a-21a9-4718-b118-90480a8b5e9c", Controller:(*bool)(0xc0024eba7a), BlockOwnerDeletion:(*bool)(0xc0024eba7b)}}}: unable to get REST mapping for cluster.x-k8s.io/v1alpha3/Machine.
```

At this point the machine exists and is running, but is on `v1beta1`, but the in-memory representation in the garbage collector still records it as being part of `v1alpha3` and can't sync the ownerReferences.

I'm going to try to replicate this on v1.27 to see if it's still an issue in more recent versions of Kubernetes.

**Comment by killianmuldoon** [comment]
Tried this with a v1.27 cluster with v1alpha4 changed to unserved on latest  - the issue is the same so this will definitely require a restart of the kube-controller-manager.

I'll open a separate issue to see how best to communicate that.

File: api/v1alpha3/machine_types.go [hunk 4]
```diff
@@ -266,6 +267,7 @@ func (m *Machine) SetConditions(conditions Conditions) {
 }
 
 // +kubebuilder:object:root=true
+// +kubebuilder:unservedversion
```
Comment (sbueringer) [inline api/v1alpha3/machine_types.go:270]: Do we need this on the list type?

File: api/v1alpha3/webhook_test.go [hunk 5]
```diff
@@ -30,6 +30,8 @@ import (
 )
 
 func TestClusterConversion(t *testing.T) {
+	t.Skipf("v1alpha3 is no longer served see: https://cluster-api.sigs.k8s.io/contributing#removal-of-v1alpha3--v1alpha4-apiversions")
```
Comment (sbueringer) [inline api/v1alpha3/webhook_test.go:33]: I would prefer deleting the tests. It's not like we're just skipping them temporarily :)

(same in other cases)

File: cmd/clusterctl/client/cluster/crd_migration_test.go [hunk 6]
```diff
@@ -184,15 +184,76 @@ func Test_CRDMigrator(t *testing.T) {
...
+						{Name: "v1beta1", Served: true},           // v1beta1 still there (required for migration)
 						// v1alpha1 is being dropped
 					},
 				},
 			},
 			wantStoredVersions: []string{"v1beta1"}, // v1alpha1 should be dropped from current CRD's stored versions
 			wantIsMigrated:     true,
 		},
+		{
+			name: "Migrate if the CR if their storage version is no longer served by the CRD",
```
Comment (sbueringer) [inline cmd/clusterctl/client/cluster/crd_migration_test.go:197]:
```suggestion
			name: "Migrate the CR if their storage version is no longer served by the CRD",
```

File: cmd/clusterctl/client/cluster/crd_migration_test.go [hunk 7]
```diff
@@ -184,15 +184,76 @@ func Test_CRDMigrator(t *testing.T) {
...
+			},
+			newCRD: &apiextensionsv1.CustomResourceDefinition{
+				ObjectMeta: metav1.ObjectMeta{Name: "foo"},
+				Spec: apiextensionsv1.CustomResourceDefinitionSpec{
+					Group: "foo",
+					Names: apiextensionsv1.CustomResourceDefinitionNames{Kind: "Foo", ListKind: "FooList"},
+					Versions: []apiextensionsv1.CustomResourceDefinitionVersion{
+						{Name: "v1", Storage: true, Served: true}, // v1 is being added
+						{Name: "v1beta1", Served: true},           // v1beta1 still there (required for migration)
+						{Name: "v1alpha1", Served: false},         // v1alpha1 is no longer being served
```
Comment (sbueringer) [inline cmd/clusterctl/client/cluster/crd_migration_test.go:250]:
```suggestion
						{Name: "v1alpha1", Served: false},         // v1alpha1 is no longer being served (required for migration)
```

File: docs/book/src/developer/providers/migrations/v1.4-to-v1.5.md [hunk 8]
```diff
@@ -17,11 +17,11 @@ maintainers of providers and consumers of our Go API.
 
 ### Deprecation
 
--
+- API version `v1alpha3` is not served anymore from v1.5 (users can enable it manually in case they are lagging behind with deprecation cycles). Important: `v1alpha3` will be removed in 1.6.
```
Comment (sbueringer) [inline docs/book/src/developer/providers/migrations/v1.4-to-v1.5.md:20]: Not sure I understand why v1alpha3 is under deprecation and v1alpha4 under removal

File: docs/book/src/user/quick-start.md [hunk 9]
```diff
@@ -83,6 +83,8 @@ a target [management cluster] on the selected [infrastructure provider].
    cat > kind-cluster-with-extramounts.yaml <<EOF
    kind: Cluster
    apiVersion: kind.x-k8s.io/v1alpha4
+   networking:
```
Comment (sbueringer) [inline docs/book/src/user/quick-start.md:86]: unrelated change

File: hack/kind-install-for-capd.sh [hunk 10]
```diff
@@ -51,6 +51,8 @@ fi
 cat <<EOF | kind create cluster --name="$KIND_CLUSTER_NAME"  --config=-
 kind: Cluster
 apiVersion: kind.x-k8s.io/v1alpha4
+networking:
```
Comment (sbueringer) [inline hack/kind-install-for-capd.sh:54]: Same

File: test/e2e/cluster_upgrade_runtimesdk_test.go [hunk 11]
```diff
@@ -26,7 +26,7 @@ import (
 	"k8s.io/utils/pointer"
 )
 
-var _ = Describe("When upgrading a workload cluster using ClusterClass with RuntimeSDK [PR-Informing] [ClusterClass]", func() {
```
Comment (sbueringer) [inline test/e2e/cluster_upgrade_runtimesdk_test.go:29]: Please double-check test/*. There seem to be a lot of unrelated changes

File: api/v1alpha3/machine_types.go (hunk 4 above)
Comment (killianmuldoon) [reply api/v1alpha3/machine_types.go]: No need - should be removed now.

File: docs/book/src/developer/providers/migrations/v1.4-to-v1.5.md (hunk 8 above)
Comment (sbueringer) [reply docs/book/src/developer/providers/migrations/v1.4-to-v1.5.md]: That would have been my general feeling :D

File: test/e2e/clusterctl_upgrade_test.go [hunk 12]
```diff
@@ -25,7 +25,8 @@ import (
 	"sigs.k8s.io/cluster-api/test/framework"
 )
 
-var _ = Describe("When testing clusterctl upgrades (v0.3=>current)", func() {
+// Note: v1.4 is the last version which serves v0.3 APIs. This is the last CAPI version that v0.3 types can be upgraded to.
+var _ = Describe("When testing clusterctl upgrades (v0.3=>v1.4) [PR-Informing]", func() {
```
Comment (sbueringer) [inline test/e2e/clusterctl_upgrade_test.go:29]: If I'm not mistaken we can just delete this test.

It is installing 0.3 and upgrading to 1.4. All based on GitHub releases. (with the exception of cluster templates)

We already have a test from v0.3 to 1.4 on the release-1.4 branch

**Review by sbueringer — COMMENTED** [review]
Last one from my side

**Review by sbueringer — COMMENTED** [review]
Last one from my side

File: test/e2e/clusterctl_upgrade_test.go (hunk 12 above)
Comment (killianmuldoon) [reply test/e2e/clusterctl_upgrade_test.go:29]: Sure - we can also get rid of the templates, right?

File: test/e2e/clusterctl_upgrade_test.go [hunk 13]
```diff
@@ -25,40 +25,6 @@ import (
 	"sigs.k8s.io/cluster-api/test/framework"
 )
 
-var _ = Describe("When testing clusterctl upgrades (v0.3=>current)", func() {
```
Comment (sbueringer) [inline test/e2e/clusterctl_upgrade_test.go:28]: My bad. Let's please also adjust our docs. Should be at least: release-tasks, book where we have the test table

File: docs/book/src/reference/jobs.md [hunk 14]
```diff
@@ -116,7 +116,7 @@ Prow Periodics:
 [periodic-cluster-api-test-main]: https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api#capi-test-main
 [periodic-cluster-api-test-mink8s-main]: https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api#capi-test-mink8s-main
 [periodic-cluster-api-e2e-main]: https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api#capi-e2e-main
-[periodic-cluster-api-e2e-upgrade-v0-3-to-main]: https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api#capi-e2e-upgrade-v0-3-to-main
+[periodic-cluster-api-e2e-upgrade-v0-4-to-main]: https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api#capi-e2e-upgrade-v0-4-to-main
 [periodic-cluster-api-e2e-upgrade-v1-0-to-main]: https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api#capi-e2e-upgrade-v1-0-to-main
```
Comment (sbueringer) [inline docs/book/src/reference/jobs.md:120]:
```suggestion
```
links are unused and those jobs don't exist anymore

File: docs/release/release-tasks.md [hunk 15]
```diff
@@ -93,26 +93,25 @@ is used for e.g. local development and e2e tests. We also modify tests so that t
 
 This comes down to changing occurrences of the old version to the new version, e.g. `v1.3` to `v1.4`:
```
Comment (sbueringer) [inline docs/release/release-tasks.md:94]: Some more:
* https://github.com/kubernetes-sigs/cluster-api/blob/187e4163a2eeabe930dda1e89f61a3651db9f0d9/test/e2e/data/infrastructure-docker/README.md#L12 (should be updated or written more generically)
* [Clusterctl upgrade test coverage](https://main.cluster-api.sigs.k8s.io/clusterctl/commands/upgrade.html#clusterctl-upgrade-test-coverage)

Comment (sbueringer) [reply docs/release/release-tasks.md:94]: I think otherwise it's fine

**Review by fabriziopandini — COMMENTED** [review]
Thanks for improving the CRD migration scripts in clusterctl, Great work! 
/lgtm
/approve
